Sources/OutlineView/TreeMap.swift

```diff
@@ -0,0 +1,223 @@
+
+/// An object representing the state of an OutlineView in which the content
+/// of the OutlineView is Identifiable. The TreeMap should be able to stay
+/// in sync with the OutlineView by reacting to data updates in the OutlineView.
+class TreeMap<D: Hashable> {
+        
+    struct Node {
+        var parentId: D?
+        var isLeaf: Bool
+        
+        // childIds can be nil if `isLeaf` == true, or if the item is not expanded.
+        // If childIds is not nil, that means the item is expanded.
+        var childIds: [D]?
+    }
+    
+    private (set) var rootData: [D] = []
+    private var directory: [D : Node] = [:]
+        
+    init() {}
+    
+    @available(macOS 10.15, *)
+    init<Data: Sequence>(
+        rootItems: [OutlineViewItem<Data>],
+        itemIsExpanded: ((OutlineViewItem<Data>) -> Bool)
+    ) where Data.Element: Identifiable, Data.Element.ID == D {
+        
+        // Add root items first
+        for item in rootItems {
+            addItem(item.value.id, isLeaf: item.children == nil, intoItem: nil, atIndex: nil)
+        }
+        
+        // Loop through all items, adding their children if they're expanded
+        var checkingItems: [(parentId: D?, item: OutlineViewItem<Data>)] = rootItems.map { (nil, $0) }
+        while !checkingItems.isEmpty {
+            let nextItem = checkingItems.popLast()!.item
+            if itemIsExpanded(nextItem),
+               let children = nextItem.children
+            {
+                let expandingChildren = children.map { ($0.id, $0.children == nil) }
+                expandItem(nextItem.id, children: expandingChildren)
+                checkingItems.append(contentsOf: children.map({ (nextItem.id, $0) }))
+            }
+        }
+    }
+    
+    /// Adds an item to the TreeMap
+    ///
+    /// - Parameters:
+    ///   - item: The ID of the item to add.
+    ///   - isLeaf: True if this item can have no children.
+    ///   - intoItem: The parent ID to insert this item into. If
+    ///   intoItem is nil, the new item is inserted at the root level.
+    ///   - atIndex: The index among other parent's children at which to
+    ///   insert this item. If no index is given, the item is added to the
+    ///   end of the array of children for the parent (or root if no
+    ///   parent is given).
+    func addItem(_ item: D, isLeaf: Bool, intoItem: D?, atIndex: Int?) {
+        
+        // Add to parent or root
+        if let intoItem {
+            // Add to children of selected item
+            if let atIndex {
+                // insert at index
+                directory[intoItem]!.childIds?.insert(item, at: atIndex)
```

Comment:
The force unwrapped dictionary lookups throughout the file slightly scare me.

Reply:
I think most of those were safe, but I agree in that I'd rather keep force-unwraps out. I cleaned those up here: https://github.com/Sameesunkaria/OutlineView/pull/10/commits/dc0a26534f53cd2629eec3e98173660986116efd

Although there is still one in `TreeMap.addItem(_:isLeaf:intoItem:atIndex:)`, it happens inside an `if` statement that checks for the existence of that force-unwrapped variable. It could be changed to use an optional-unwrap if you still think that's best.

Reply:
Thank you for taking care of it. While some force unwraps might be safe, I think of avoiding them as a modeling exercise. In the one remaining case you can modify the stored value and re-assign it.

```swift
if let intoItem,
  var fullIntoItem = directory[intoItem],
  case var .expanded(intoChildren) = fullIntoItem.state
{
  // Add to children of selected item
  if let atIndex {
    // insert at index
    intoChildren.insert(item, at: atIndex)
  } else if atIndex == nil {
    // append to end
    intoChildren.append(item)
  }

  fullIntoItem.state = .expanded(children: intoChildren)
  directory[intoItem] = fullIntoItem
}
```

Reply:
Good call! Fixed in https://github.com/Sameesunkaria/OutlineView/pull/10/commits/b00ac09c8a614dcb7e0146a63f3a8fb55c39dbf8